Iceberg table partition creation fails due to column name conflict with  Trino's generated partitioning column name

Sample DDL:
``` sql
create table lakehouse."default".mattest (
	starttime timestamp,
	starttime_day varchar
) with (partitioning = array['day(starttime)']);
```
Relevant stacktrace

```log
Query failed (#20240610_193515_00091_n2dt5): Unable to parse partitioning value: Cannot create partition from name that exists in schema: starttime_day
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:133)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeStatement(SQLQueryJob.java:601)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.lambda$2(SQLQueryJob.java:504)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:190)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeSingleQuery(SQLQueryJob.java:511)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.extractData(SQLQueryJob.java:963)
	at org.jkiss.dbeaver.ui.editors.sql.SQLEditor$QueryResultsContainer.readData(SQLEditor.java:4134)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:123)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:190)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:121)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:5148)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:114)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.sql.SQLException: Query failed (#20240610_193515_00091_n2dt5): Unable to parse partitioning value: Cannot create partition from name that exists in schema: starttime_day
	at io.trino.jdbc.AbstractTrinoResultSet.resultsException(AbstractTrinoResultSet.java:1937)
	at io.trino.jdbc.TrinoResultSet$ResultsPageIterator.computeNext(TrinoResultSet.java:294)
	at io.trino.jdbc.TrinoResultSet$ResultsPageIterator.computeNext(TrinoResultSet.java:254)
	at io.trino.jdbc.$internal.guava.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:145)
	at io.trino.jdbc.$internal.guava.collect.AbstractIterator.hasNext(AbstractIterator.java:140)
	at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Unknown Source)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$0(Unknown Source)
	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(Unknown Source)
	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(Unknown Source)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(Unknown Source)
	at java.base/java.util.Spliterators$1Adapter.hasNext(Unknown Source)
	at io.trino.jdbc.TrinoResultSet$AsyncIterator.lambda$new$1(TrinoResultSet.java:179)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.trino.spi.TrinoException: Unable to parse partitioning value: Cannot create partition from name that exists in schema: starttime_day
	at io.trino.plugin.iceberg.PartitionFields.parsePartitionFields(PartitionFields.java:69)
	at io.trino.plugin.iceberg.IcebergMetadata.getNewTableLayout(IcebergMetadata.java:917)
	at io.trino.plugin.iceberg.IcebergMetadata.createTable(IcebergMetadata.java:884)
	at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.createTable(ClassLoaderSafeConnectorMetadata.java:457)
	at io.trino.tracing.TracingConnectorMetadata.createTable(TracingConnectorMetadata.java:394)
	at io.trino.metadata.MetadataManager.createTable(MetadataManager.java:818)
	at io.trino.tracing.TracingMetadata.createTable(TracingMetadata.java:409)
	at io.trino.execution.CreateTableTask.internalExecute(CreateTableTask.java:297)
	at io.trino.execution.CreateTableTask.execute(CreateTableTask.java:126)
	at io.trino.execution.CreateTableTask.execute(CreateTableTask.java:92)
	at io.trino.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:146)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:272)
	at io.trino.dispatcher.LocalDispatchQuery.startExecution(LocalDispatchQuery.java:145)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$waitForMinimumWorkers$2(LocalDispatchQuery.java:129)
	at io.airlift.concurrent.MoreFutures.lambda$addSuccessCallback$12(MoreFutures.java:570)
	at io.airlift.concurrent.MoreFutures$3.onSuccess(MoreFutures.java:545)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1137)
	at io.trino.$gen.Trino_443____20240604_192937_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalArgumentException: Cannot create partition from name that exists in schema: starttime_day
	at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:218)
	at org.apache.iceberg.PartitionSpec$Builder.checkAndAddPartitionName(PartitionSpec.java:380)
	at org.apache.iceberg.PartitionSpec$Builder.checkAndAddPartitionName(PartitionSpec.java:358)
	at org.apache.iceberg.PartitionSpec$Builder.day(PartitionSpec.java:461)
	at org.apache.iceberg.PartitionSpec$Builder.day(PartitionSpec.java:471)
	at io.trino.plugin.iceberg.PartitionFields.lambda$parsePartitionField$3(PartitionFields.java:78)
	at io.trino.plugin.iceberg.PartitionFields.tryMatch(PartitionFields.java:110)
	at io.trino.plugin.iceberg.PartitionFields.parsePartitionField(PartitionFields.java:78)
	at io.trino.plugin.iceberg.PartitionFields.parsePartitionFields(PartitionFields.java:64)
	... 20 more

```

I think having a way to alias/name the partitioning column can help resolve this. An example on [spark side](https://iceberg.apache.org/docs/nightly/spark-ddl/#alter-table-add-partition-field)

Slack thread, https://trinodb.slack.com/archives/C0305TQ05KL/p1717577005413109.

Tagging @findinpath

--- a/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
+++ b/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java
@@ -1922,9 +1922,8 @@ public class IcebergMetadata
             difference(existingPartitionFields, partitionFields).stream()
                     .map(PartitionField::name)
                     .forEach(updatePartitionSpec::removeField);
-            difference(partitionFields, existingPartitionFields).stream()
-                    .map(partitionField -> toIcebergTerm(schema, partitionField))
-                    .forEach(updatePartitionSpec::addField);
+            difference(partitionFields, existingPartitionFields)
+                    .forEach(partitionField -> updatePartitionSpec.addField(partitionField.name(), toIcebergTerm(schema, partitionField)));
         }
 
         try {

--- a/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionFields.java
+++ b/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/PartitionFields.java
@@ -61,7 +61,7 @@ public final class PartitionFields
         try {
             PartitionSpec.Builder builder = PartitionSpec.builderFor(schema);
             for (String field : fields) {
-                parsePartitionField(builder, field);
+                parsePartitionFields(schema, fields, builder, field);
             }
             return builder.build();
         }
@@ -70,16 +70,59 @@ public final class PartitionFields
         }
     }
 
-    public static void parsePartitionField(PartitionSpec.Builder builder, String field)
+    private static void parsePartitionFields(Schema schema, List<String> fields, PartitionSpec.Builder builder, String field)
     {
-        boolean matched = tryMatch(field, IDENTITY_PATTERN, match -> builder.identity(fromIdentifierToColumn(match.group()))) ||
-                tryMatch(field, YEAR_PATTERN, match -> builder.year(fromIdentifierToColumn(match.group(1)))) ||
-                tryMatch(field, MONTH_PATTERN, match -> builder.month(fromIdentifierToColumn(match.group(1)))) ||
-                tryMatch(field, DAY_PATTERN, match -> builder.day(fromIdentifierToColumn(match.group(1)))) ||
-                tryMatch(field, HOUR_PATTERN, match -> builder.hour(fromIdentifierToColumn(match.group(1)))) ||
-                tryMatch(field, BUCKET_PATTERN, match -> builder.bucket(fromIdentifierToColumn(match.group(1)), parseInt(match.group(2)))) ||
-                tryMatch(field, TRUNCATE_PATTERN, match -> builder.truncate(fromIdentifierToColumn(match.group(1)), parseInt(match.group(2)))) ||
-                tryMatch(field, VOID_PATTERN, match -> builder.alwaysNull(fromIdentifierToColumn(match.group(1))));
+        for (int i = 1; i < schema.columns().size() + fields.size(); i++) {
+            try {
+                parsePartitionField(builder, field, i == 1 ? "" : "_" + i);
+                return;
+            }
+            catch (IllegalArgumentException e) {
+                if (e.getMessage().contains("Cannot create partition from name that exists in schema")
+                        || e.getMessage().contains("Cannot create identity partition sourced from different field in schema")) {
+                    continue;
+                }
+                throw e;
+            }
+        }
+        throw new IllegalArgumentException("Cannot resolve partition field: " + field);
+    }
+
+    public static void parsePartitionField(PartitionSpec.Builder builder, String field, String suffix)
+    {
+        boolean matched =
+                tryMatch(field, IDENTITY_PATTERN, match -> {
+                    // identity doesn't allow specifying an alias
+                    builder.identity(fromIdentifierToColumn(match.group()));
+                }) ||
+                tryMatch(field, YEAR_PATTERN, match -> {
+                    String column = fromIdentifierToColumn(match.group(1));
+                    builder.year(column, column + "_year" + suffix);
+                }) ||
+                tryMatch(field, MONTH_PATTERN, match -> {
+                    String column = fromIdentifierToColumn(match.group(1));
+                    builder.month(column, column + "_month" + suffix);
+                }) ||
+                tryMatch(field, DAY_PATTERN, match -> {
+                    String column = fromIdentifierToColumn(match.group(1));
+                    builder.day(column, column + "_day" + suffix);
+                }) ||
+                tryMatch(field, HOUR_PATTERN, match -> {
+                    String column = fromIdentifierToColumn(match.group(1));
+                    builder.hour(column, column + "_hour" + suffix);
+                }) ||
+                tryMatch(field, BUCKET_PATTERN, match -> {
+                    String column = fromIdentifierToColumn(match.group(1));
+                    builder.bucket(column, parseInt(match.group(2)), column + "_bucket" + suffix);
+                }) ||
+                tryMatch(field, TRUNCATE_PATTERN, match -> {
+                    String column = fromIdentifierToColumn(match.group(1));
+                    builder.truncate(column, parseInt(match.group(2)), column + "_trunc" + suffix);
+                }) ||
+                tryMatch(field, VOID_PATTERN, match -> {
+                    String column = fromIdentifierToColumn(match.group(1));
+                    builder.alwaysNull(column, column + "_null" + suffix);
+                });
         if (!matched) {
             throw new IllegalArgumentException("Invalid partition field declaration: " + field);
         }

--- a/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java
+++ b/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java
@@ -1316,6 +1316,29 @@ public abstract class BaseIcebergConnectorTest
     }
 
     @Test
+    public void testPartitionColumnNameConflict()
+    {
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_conflict_partition", "(ts timestamp, ts_day int) WITH (partitioning = ARRAY['day(ts)'])")) {
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (TIMESTAMP '2021-07-24 03:43:57.987654', 1)", 1);
+
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES (TIMESTAMP '2021-07-24 03:43:57.987654', 1)");
+            assertThat(query("SELECT partition.ts_day_2 FROM \"" + table.getName() + "$partitions\""))
+                    .matches("VALUES DATE '2021-07-24'");
+        }
+
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_conflict_partition", "(ts timestamp, ts_day int)")) {
+            assertUpdate("ALTER TABLE " + table.getName() + " SET PROPERTIES partitioning = ARRAY['day(ts)']");
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (TIMESTAMP '2021-07-24 03:43:57.987654', 1)", 1);
+
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES (TIMESTAMP '2021-07-24 03:43:57.987654', 1)");
+            assertThat(query("SELECT partition.ts_day_2 FROM \"" + table.getName() + "$partitions\""))
+                    .matches("VALUES DATE '2021-07-24'");
+        }
+    }
+
+    @Test
     public void testSortByAllTypes()
     {
         String tableName = "test_sort_by_all_types_" + randomNameSuffix();

--- a/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestPartitionFields.java
+++ b/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestPartitionFields.java
@@ -13,9 +13,12 @@
  */
 package io.trino.plugin.iceberg;
 
+import com.google.common.collect.ImmutableList;
+import org.apache.iceberg.PartitionField;
 import org.apache.iceberg.PartitionSpec;
 import org.apache.iceberg.Schema;
 import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.types.Type;
 import org.apache.iceberg.types.Types;
 import org.apache.iceberg.types.Types.DoubleType;
 import org.apache.iceberg.types.Types.ListType;
@@ -25,10 +28,12 @@ import org.apache.iceberg.types.Types.StringType;
 import org.apache.iceberg.types.Types.TimestampType;
 import org.junit.jupiter.api.Test;
 
+import java.util.List;
 import java.util.function.Consumer;
 
 import static com.google.common.collect.Iterables.getOnlyElement;
 import static io.trino.plugin.iceberg.PartitionFields.parsePartitionField;
+import static io.trino.plugin.iceberg.PartitionFields.parsePartitionFields;
 import static io.trino.plugin.iceberg.PartitionFields.toPartitionFields;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
@@ -96,6 +101,35 @@ public class TestPartitionFields
         assertInvalid("\"nested.list\"", "Cannot partition by non-primitive source field: list<string>");
     }
 
+    @Test
+    public void testConflicts()
+    {
+        assertParseName(List.of("col", "col_year"), TimestampType.withZone(), List.of("year(col)"), List.of("col_year_2"));
+        assertParseName(List.of("col", "col_month"), TimestampType.withZone(), List.of("month(col)"), List.of("col_month_2"));
+        assertParseName(List.of("col", "col_day"), TimestampType.withZone(), List.of("day(col)"), List.of("col_day_2"));
+        assertParseName(List.of("col", "col_hour"), TimestampType.withZone(), List.of("hour(col)"), List.of("col_hour_2"));
+        assertParseName(List.of("col", "col_bucket"), TimestampType.withZone(), List.of("bucket(col,10)"), List.of("col_bucket_2"));
+        assertParseName(List.of("col", "col_trunc"), StringType.get(), List.of("truncate(col,10)"), List.of("col_trunc_2"));
+        assertParseName(List.of("col", "col_null"), TimestampType.withZone(), List.of("void(col)"), List.of("col_null_2"));
+
+        assertParseName(List.of("col", "col_year", "col_year_2"), TimestampType.withZone(), List.of("year(col)"), List.of("col_year_3"));
+        assertParseName(List.of("col", "col_year", "col_year_3"), TimestampType.withZone(), List.of("year(col)"), List.of("col_year_2"));
+
+        assertParseName(List.of("col", "col_year", "col_year_2"), TimestampType.withZone(), List.of("year(col)", "col_year_2"), List.of("col_year_3", "col_year_2"));
+    }
+
+    private static void assertParseName(List<String> columnNames, Type type, List<String> partitions, List<String> expected)
+    {
+        ImmutableList.Builder<NestedField> columns = ImmutableList.builderWithExpectedSize(columnNames.size());
+        int i = 1;
+        for (String name : columnNames) {
+            columns.add(NestedField.required(i++, name, type));
+        }
+        PartitionSpec spec = parsePartitionFields(new Schema(columns.build()), partitions);
+        assertThat(spec.fields()).extracting(PartitionField::name)
+                .containsExactlyElementsOf(expected);
+    }
+
     private static void assertParse(String value, PartitionSpec expected, String canonicalRepresentation)
     {
         assertThat(expected.fields().size()).isEqualTo(1);
@@ -120,7 +154,7 @@ public class TestPartitionFields
 
     private static PartitionSpec parseField(String value)
     {
-        return partitionSpec(builder -> parsePartitionField(builder, value));
+        return partitionSpec(builder -> parsePartitionField(builder, value, ""));
     }
 
     private static PartitionSpec partitionSpec(Consumer<PartitionSpec.Builder> consumer)

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
@@ -2454,6 +2454,24 @@ public class TestIcebergSparkCompatibility
     }
 
     @Test(groups = {ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testPartitionColumnNameConflict()
+    {
+        String baseTableName = "test_conflict_partition" + randomNameSuffix();
+        String trinoTableName = trinoTableName(baseTableName);
+        String sparkTableName = sparkTableName(baseTableName);
+
+        onTrino().executeQuery("CREATE TABLE " + trinoTableName + "(ts timestamp, ts_day int) WITH (partitioning = ARRAY['day(ts)'])");
+        onTrino().executeQuery("INSERT INTO " + trinoTableName + " VALUES (TIMESTAMP '2021-07-24 03:43:57.987654', 1)");
+
+        assertThat(onSpark().executeQuery("SELECT * FROM " + sparkTableName))
+                .containsOnly(row(Timestamp.valueOf("2021-07-24 03:43:57.987654"), 1));
+        assertThat(onSpark().executeQuery("SELECT partition['ts_day_2'] FROM " + sparkTableName + ".partitions"))
+                .containsOnly(row(Date.valueOf("2021-07-24")));
+
+        onTrino().executeQuery("DROP TABLE " + trinoTableName);
+    }
+
+    @Test(groups = {ICEBERG, PROFILE_SPECIFIC_TESTS})
     public void testHandlingPartitionSchemaEvolutionInPartitionMetadata()
     {
         String baseTableName = "test_handling_partition_schema_evolution_" + randomNameSuffix();

